docs/topics/using-builders-with-builder-inference.md

```diff
@@ -0,0 +1,277 @@
+receiver.
+
+### Contributing to builder inference results
+
+Builder inference can collect different flavors of type information that contributes to the analysis result.
+It considers:
+* Calling methods on a lambda's receiver that use a type parameter's type:
+  ```kotlin
+  val result = buildList {
+      // Type argument is inferred into String based on the passed "value" argument
+      add("value")
+  } // result has the List<String> type inferred
+  ```
+* Specifying the expected type for calls which return type parameter's type:
+  ```kotlin
+  val result = buildList {
+      // Type argument is inferred into Float based on the expected type
+      val x: Float = get(0)
+  } // result has the List<Float> type
+  ```
+  ```kotlin
+  class Foo<T> {
+      val items = mutableListOf<T>()
+  }
+
+  fun <K> myBuilder(builder: Foo<K>.() -> Unit): Foo<K> = Foo<K>().apply(builder)
+
+  fun main() {
+      val result = myBuilder {
+          val x: List<CharSequence> = items
+          // ...
+      } // result has the Foo<CharSequence> type
+  }
+  ```
+* Passing postponed type variables' types into methods which expect concrete types:
+  ```kotlin
+  fun takeMyLong(x: Long) { ... }
+
+  fun String.isMoreThat3() = length > 3
+
+  fun takeListOfStrings(x: List<String>) { ... }
+
+  fun main() {
+      val result1 = buildList {
+          val x = get(0)
+          takeMyLong(x)
+      } // result1 has the List<Long> type
+
+      val result2 = buildList {
+          val x = get(0)
+          val isLong = x.isMoreThat3()
+      // ...
+      } // result2 has the List<String> type
+  
+      val result3 = buildList {
+          takeListOfStrings(this)
+      } // result3 has the List<String> type
+  }
+  ```
+* Taking a callable reference to lambda receiver's member:
+  ```kotlin
+  fun main() {
+      val result = buildList {
+          val x: KFunction1<Int, Float> = ::get
+      } // result has the List<Float> type
+  }
+  ```
+  ```kotlin
+  fun takeFunction(x: KFunction1<Int, Float>) { ... }
+
+  fun main() {
+      val result = buildList {
+          takeFunction(::get)
+      } // result has the List<Float> type
+  }
+  ```
+* Using nested interdependent builder inference calls.
+
+At the end of the analysis, builder inference considers all collected type information and tries to merge it into 
+the resulting type. See the example.
+
+```kotlin
+val result = buildList { // Inferring postponed type variable E
+    // Considering E is Number or a subtype of Number
+    val n: Number? = getOrNull(0)
+    // Considering E is Int or a supertype of Int
+    add(1)
+    // E gets inferred into Int
+} // result has the List<Int> type
+```
+
+So the resulting type is the most specific type that corresponds to the type information collected during the analysis.
+If given type information is contradictory and cannot be merged, the compiler reports an error.
```

Comment:
@petukhovv This is the additional info you requested in DM. I tried to translate terms like upper or lower bounds into sub- and supertypes.

Reply:
Looks good to me